cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -158,24 +164,36 @@ private boolean doesLeaderRemoveViolateOptimizedGoal(Replica replicaToBeRemoved)
   @Override
   protected void initGoalState(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
       throws OptimizationFailureException {
-    _mustHaveTopicLeadersPerBroker = Collections.unmodifiableSet(
+    Set<String> mustHaveTopicLeadersPerBroker = Collections.unmodifiableSet(
         Utils.getTopicNamesMatchedWithPattern(_balancingConstraint.topicsWithMinLeadersPerBrokerPattern(), clusterModel::topics));
-    if (_mustHaveTopicLeadersPerBroker.isEmpty()) {
+    _mustHaveTopicMinLeadersPerBroker = new HashMap<>();
+    // populate min leaders per broker for each topic when dynamic computation is configured
+    if (mustHaveTopicLeadersPerBroker.isEmpty()) {
       return;
     }
+    Map<String, Integer> numLeadersByTopicNames = clusterModel.numLeadersPerTopic(mustHaveTopicLeadersPerBroker);
+    int eligibleBrokersForLeadership = eligibleBrokersForLeadership(clusterModel, optimizationOptions).size();
```

Comment:
Nit: `#validateEnoughLeaderToDistribute(ClusterModel, OptimizationOptions)` function, which is only called from `#initGoalState`, seems to recompute the same values for
1. `numLeadersByTopicNames` and 
2. `eligibleBrokersForLeadership(clusterModel, optimizationOptions)`. 
 
Would it be possible to change the `#validateEnoughLeaderToDistribute(ClusterModel, OptimizationOptions)` to something like `#validateEnoughLeaderToDistribute(Map<String, Integer> numLeadersByTopicNames, Set<Broker> eligibleBrokersForLeadership)` and pass the parameters computed here?

Reply:
Agree - this has been implemented